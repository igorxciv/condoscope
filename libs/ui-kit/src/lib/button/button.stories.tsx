import React, { FC } from 'react'
import { select, text } from '@storybook/addon-knobs'
import { Button } from './button'

export default {
  component: Button,
  title: 'Button',
}

export const filled: FC = () => {
  return (
    <Button
      size={select('Size', {
        Tiny: 'tiny',
        Small: 'small',
        Medium: 'medium',
        Large: 'large',
        Giant: 'giant',
      }, 'medium')}
      color={select('Color', {
        Basic: 'basic',
        Primary: 'primary',
        Success: 'success',
        Info: 'info',
        Warning: 'warning',
        Danger: 'danger',
      }, 'basic')}
    >
      {text('Label', 'Button')}
    </Button>
  )
}
