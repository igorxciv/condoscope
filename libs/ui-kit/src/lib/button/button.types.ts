export type ButtonSize = 'giant' | 'large' | 'medium' | 'small' | 'tiny'
export type ButtonColor = 'basic' | 'primary' | 'success' | 'info' | 'warning' | 'danger'
export type ButtonAppearance = 'filled' | 'outline' | 'ghost'

export type ButtonProps = {
  size?: ButtonSize,
  appearance?: ButtonAppearance,
  color?: ButtonColor,
}
