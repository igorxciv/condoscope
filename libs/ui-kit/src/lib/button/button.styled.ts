import styled, { css } from 'styled-components'
import { ButtonProps } from './button.types'
import { propertyAnimation } from '../__design/utils'

const filled = css<ButtonProps>`

  ${({ theme, status }) => {
    const textColor = status === 'basic' ? theme['button-filled-basic-color'] : theme['button-filled-colorized-color']

    return css`
      color: ${textColor};
      background-color: ${theme[`button-filled-${status}-background`]};

      &:hover {
        background-color: ${theme[`button-filled-${status}-hover-background`]};
      }
  `}
}`

const outline = css<ButtonProps>`
  ${({ theme, status }) => css`
    border: 1px solid ${theme[`button-outline-${status}-border`]};
    color: ${theme[`button-outline-${status}-color`]};
    background-color: ${theme[`button-outline-${status}-background`]};

    &:hover {
      background-color: ${theme[`button-outline-${status}-hover-background`]};
    }
  `}
`

const ghost = css<ButtonProps>`
  ${({ theme, status }) => css`
    color: ${theme[`button-ghost-${status}-color`]};
    background-color: ${theme['button-ghost-background']};

    &:hover {
      background-color: ${theme['button-ghost-hover-background']};
    }
  `}
`

const appearances = { filled, outline, ghost }

const ButtonStyle = css<ButtonProps>`
  ${({ theme, appearance, size }) => css`
    border-radius: ${theme['button-radius']};
    border: none;
    cursor: pointer;
    text-transform: uppercase;

    font-family: ${theme['font-family-primary']};
    font-weight: ${theme['font-weight-bold']};
    font-size: ${theme[`button-${size}-font-size`]};

    padding: ${theme[`button-${size}-padding`]};

    line-height: ${theme[`button-${size}-line-height`]};

    ${propertyAnimation(['background-color'])}

    ${appearances[appearance]}
  `}
`

export const ButtonStyled = styled.button<ButtonProps>`
  ${ButtonStyle}
`
