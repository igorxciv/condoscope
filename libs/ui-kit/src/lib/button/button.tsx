import React, { FC, PropsWithChildren } from 'react'
import { ButtonStyled } from './button.styled'
import type { ButtonProps } from './button.types'

export const Button: FC<ButtonProps> = (props: PropsWithChildren<ButtonProps>) => {
  const { children } = props

  return (
    <ButtonStyled {...props}>
      {children}
    </ButtonStyled>
  )
}

Button.defaultProps = {
  appearance: 'filled',
  status: 'basic',
  size: 'medium',
}
