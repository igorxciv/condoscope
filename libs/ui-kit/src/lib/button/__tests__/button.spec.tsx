import React from 'react'
import renderer, { ReactTestRenderer } from 'react-test-renderer'
import { Button } from '../button'
import 'jest-styled-components'

describe('Button', () => {
  let component: ReactTestRenderer

  beforeEach(() => {
    component = renderer.create(
      <Button>
        Button
      </Button>,
    )
  })

  describe('Filled', () => {
    beforeEach(() => {
      component = renderer.create(
        <Button appearance="filled">
          Button
        </Button>,
      )
    })

    it('renders successfully', () => {
      expect(component.toJSON()).toMatchSnapshot()
    })
  })

  describe('Outline', () => {
    beforeEach(() => {
      component = renderer.create(
        <Button appearance="outline">
          Button
        </Button>,
      )
    })

    it('renders successfully', () => {
      expect(component.toJSON()).toMatchSnapshot()
    })
  })

  describe('Ghost', () => {
    beforeEach(() => {
      component = renderer.create(
        <Button appearance="ghost">
          Button
        </Button>,
      )
    })

    it('renders successfully', () => {
      expect(component.toJSON()).toMatchSnapshot()
    })
  })
})
