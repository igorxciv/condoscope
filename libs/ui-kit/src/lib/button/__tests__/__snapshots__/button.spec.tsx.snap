// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Filled renders successfully 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  text-transform: uppercase;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transition-property: background-color border-color;
  transition-property: background-color border-color;
  -webkit-transition-timing-function: ease-in;
  transition-timing-function: ease-in;
  background-color: var(--background-color);
  border: 0.0625rem solid var(--border-color);
}

.c0:focus {
  outline: none;
}

<button
  className="c0"
  size="medium"
>
  Button
</button>
`;

exports[`Button Ghost renders successfully 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  text-transform: uppercase;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transition-property: background-color border-color;
  transition-property: background-color border-color;
  -webkit-transition-timing-function: ease-in;
  transition-timing-function: ease-in;
  border: 0.0625rem solid;
}

.c0:focus {
  outline: none;
}

<button
  className="c0"
  size="medium"
>
  Button
</button>
`;

exports[`Button Outline renders successfully 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  text-transform: uppercase;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transition-property: background-color border-color;
  transition-property: background-color border-color;
  -webkit-transition-timing-function: ease-in;
  transition-timing-function: ease-in;
  border: 0.0625rem solid;
  background-color: var(--background-color);
}

.c0:focus {
  outline: none;
}

<button
  className="c0"
  size="medium"
>
  Button
</button>
`;
